### PR TITLE
[Hub Generated] Review request for Microsoft.BotService to add version preview/2021-05-01

### PR DESCRIPTION
**Pull request by swagatmishra2007**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/swagatmishra2007/azure-rest-api-specs/dev-botservice-Microsoft.BotService-2021-05-01...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
 We are targeting to be in public preview by July for private links (which are the new api's added in this new APIVersion
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  The swagger would be published (as public preview) in june. We expect to go to a stable version a month after that
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by swagatmishra2007** [comment]
Changes in this PR-
- Add support for operationresults, so that we can support async ARM api's
- Add support for private links

**Comment by mentat9** [comment]
Please fill out the contribution checklist.

**Comment by swagatmishra2007** [comment]
@mentat9  i have added the contribution checklist (the markdown wasnt rendering earlier for some reason). I have some CI failures, but they seem to be present in api's present in previous versions of the swagger

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1178]: >404 [](http://example.com/codeflow?start=11&length=3)

It's common to let this default, your choice.

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1185]: >default [](http://example.com/codeflow?start=11&length=7)

Add x-ms-error-response here as well.

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1192]: >"deprecated": false, [](http://example.com/codeflow?start=8&length=20)

Why is this here?

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1229]: >default [](http://example.com/codeflow?start=11&length=7)

x-ms-error-response

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1323]: >200 [](http://example.com/codeflow?start=11&length=3)

This should be 201. 200 is also OK to indicate the PUT didn't do anything (no change).

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:2815]: >startTime [](http://example.com/codeflow?start=9&length=9)

Add format: date-time

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [inline specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:2822]: Is this leftover? You need to schematize whatever other properties you want. They shouldn't be in a nested property bag.

**Comment by mentat9** [comment]
There are still several unaddressed comments. You aren't required to implement non-blocking suggestions but please reply to everything so I know you've seen and considered it.

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (swagatmishra2007) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1192]: removed. copy-paste mistake

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (swagatmishra2007) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:2815]: done

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (swagatmishra2007) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1178]: removed to let it default

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:2822]: In that case, please remove it. You'll need a new API version to add new properties anyway, so add whatever you need when you create your next API version.

**Comment by swagatmishra2007** [comment]
@mentat9 thank you for the review comments. I have answered the comments and pushed my changes in to address other feedback. Please take a look

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:1318]: Got it.

File: specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json
Comment (mentat9) [reply specification/botservice/resource-manager/Microsoft.BotService/preview/2021-05-01/botservice.json:2821]: > In that case, please remove it. You'll need a new API version to add new properties anyway, so add whatever you need when you create your next API version.

You may have missed this one. Strongly recommend only defining objects that you need.

**Comment by swagatmishra2007** [comment]
@mentat9 I have addressed the last comment (removed the property bag on operation results).
Question: I intend this new api version to be a preview, based on the understanding that there might be breaking changes to it. Is checking it under the preview folder of swagger enough or do i need to update the api version name to -preview?
(Our customers depending on this work understand that this is preview work and subject to change. Mostly , it ll be used on specific subscriptions via feature flags only)

**Comment by mentat9** [comment]
> @mentat9 I have addressed the last comment (removed the property bag on operation results).
> Question: I intend this new api version to be a preview, based on the understanding that there might be breaking changes to it. Is checking it under the preview folder of swagger enough or do i need to update the api version name to -preview?
> (Our customers depending on this work understand that this is preview work and subject to change. Mostly , it ll be used on specific subscriptions via feature flags only)

I'm not sure whether it's exactly required, but preview API versions should always be named with a -preview suffix. I'll sign off for ARM review: you can add another commit to add -preview to the name and move the files (thank-you for waiting until ARM review finished).

Going forward, you will need to add a new API version any time you add new properties, since it's a breaking change to the existing API version and clients wouldn't have a way to know whether the property is supposed to be there or not. You will only need breaking change review if you want to remove properties, but if you are still in private preview and have clear communication with all customers using the API, it shouldn't block you. This is the breaking change policy doc: https://aka.ms/azbreakingchangespolicy.

**Comment by swagatmishra2007** [comment]
@mentat9 Thanks for the review and the signoff! i have pushed a commit to change the api-version to have a -preview suffix.
Can you please merge my PR?

**Comment by mentat9** [comment]
> @mentat9 Thanks for the review and the signoff! i have pushed a commit to change the api-version to have a -preview suffix.
> Can you please merge my PR?

I'm not involved in merging. ARM review only reviews the REST APIs. The SDK team owns downstream steps (SDK review and generation, linter checks, merging).

**Comment by swagatmishra2007** [comment]
@lirenhe can you please help merge the PR? (all errors in the pipelines are from a previous version of the API that i forked off and unrelated to the newly added API's)

**Review by lirenhe — COMMENTED** [review]
@swagatmishra2007, for the lint rule error, it seems those are introduced with the latest change, could you double confirm?
And i also found a couple of breaking changes compared with previous version (stable/2021-03-01).  As breaking change cross API version need breaking change review board approval, could you follow the process to do that?

**Comment by lirenhe** [comment]
Could you also check the model validation errors? seems the example provided are not consistent with the swagger definition

**Comment by swagatmishra2007** [comment]
@lirenhe  i have fixed the new lint and model validation errors.
Regarding breaking changes, this PR only contains new API's related to private links and async operations. I did a diff between the swagger file and the swagger file for the last stable version (2021-03-01) and only see new API's being added and no breaking changes.
The pipeline also indicates that there are no breaking changes from the last stable version. 
Do you see any particular breaking changes that i am not seeing?

**Comment by swagatmishra2007** [comment]
@lirenhe  can you please take a look at it? This PR is blocking further development for private links and we have deadlines to meet for this feature work

**Comment by raych1** [comment]
@swagatmishra2007 , `model validation` failed due to duplicated model name of `Resource` in `botservice.json` and `types.json` from common-type folder. Can you reference the common `Resource` model from common-type folder since this is a new API version?

**Comment by lirenhe** [comment]
No breaking change compared with the latest version.

**Review by lirenhe — APPROVED** [review]

**Pull request merged**
